darts/models/forecasting/forecasting_model.py

```diff
@@ -834,34 +907,36 @@ def historical_forecasts(
 
             # iterate and forecast
             for _counter, pred_time in enumerate(iterator):
-                # build the training series
-                if min_timestamp > series_.time_index[0]:
+                # build the training series, as if retrain is True to break circular dependency
+                if min_timestamp_train > series_.time_index[0]:
                     train_series = series_.drop_before(
-                        min_timestamp - 1 * series_.freq
+                        min_timestamp_train - 1 * series_.freq
                     ).drop_after(pred_time)
                 else:
                     train_series = series_.drop_after(pred_time)
 
                 if train_length and len(train_series) > train_length:
                     train_series = train_series[-train_length:]
 
-                if retrain_func(
-                    counter=_counter,
-                    pred_time=pred_time,
-                    train_series=train_series,
-                    past_covariates=past_covariates_.drop_after(pred_time)
-                    if past_covariates_
-                    and ("past_covariates" in retrain_func_signature)
-                    else None,
-                    future_covariates=future_covariates_.drop_after(
-                        min(
-                            pred_time + train_series.freq * forecast_horizon,
-                            series_.end_time(),
+                # retrain_func processes the series that would be used for training
+                if (
+                    retrain_func(
+                        counter=_counter,
+                        pred_time=pred_time,
+                        train_series=train_series,
+                        past_covariates=past_covariates_.drop_after(pred_time)
```

Comment:
the same here. If we really want to drop, then we should give the correct past/future covariate time spans given by the model lags.
Otherwise, I think we can simply pass the entire past/future covariates, as the user already has the pred time, and train series up until the forecasting point.
(it can take some time to create time series with drop_after, especially in a loop like historical forecasts)

Reply:
I agree with the idea of passing the entire past/future covariates and have the user leverage the `predict_time` argument of `retrain_func` to process everything.